### PR TITLE
Issue #371, plus other enhancements/cleanup to PersonaCard

### PR DESCRIPTION
**Pull request by manndave**
Updates to PersonaCard. Bug #371 plus:

- Order of actions `<li>` and detail sections in the HTML are now irrelevant. Action has an attribute 'data-datailsTargetId' which must match the id of the target `<ul>`
- Whichever action has the 'is-active' attribute will be set to visible on initial load
- JS no longer tied to specific class names in CSS for which action/details to show (no need to change CSS if changing actions)
- Made sure it works with multiple cards on a page (so long as IDs are unique)
- Cleaned up CSS, JS and HTML
- Added support for actions to navigate to new pages via data-actionUrl attribute (i.e. for "View Profile" action.)

### CODE REVIEW

File: src/components/PersonaCard/PersonaCard.html [hunk 1]
```diff
@@ -17,31 +17,39 @@
     </div>
   </div>
   <ul class="ms-PersonaCard-actions">
-    <li id="chat" class="ms-PersonaCard-action is-active"><i class="ms-Icon ms-Icon--chat"></i></li>
-    <li id="phone" class="ms-PersonaCard-action"><i class="ms-Icon ms-Icon--phone"></i></li>
-    <li id="video" class="ms-PersonaCard-action"><i class="ms-Icon ms-Icon--video"></i></li>
-    <li id="mail" class="ms-PersonaCard-action"><i class="ms-Icon ms-Icon--mail"></i></li>
-    <li class="ms-PersonaCard-overflow" alt="View profile in Delve" title="View profile in Delve">View profile</li>
-    <li id="org" class="ms-PersonaCard-action ms-PersonaCard-orgChart"><i class="ms-Icon ms-Icon--org"></i></li>
+    <li id="phoneAction" class="ms-PersonaCard-action" data-detailsTargetId="detail-2"><i class="ms-Icon ms-Icon--phone"></i></li>
```
Comment (mikewheaton) [inline src/components/PersonaCard/PersonaCard.html]: These IDs (`phoneAction`, `chatAction`, etc.) should not be necessary, right? I removed them and it works fine. Let's keep everything generic so that it's clear that these tabs can be modified.

File: src/components/PersonaCard/PersonaCard.html [hunk 2]
```diff
@@ -17,31 +17,39 @@
...
+   </ul>
+   <ul id="detail-4" class="ms-PersonaCard-details">
+      <li class="ms-PersonaCard-actionDetails">
         <div class="ms-PersonaCard-detailLine"><span class="ms-PersonaCard-detailLabel">Personal:</span> <a class="ms-Link" href="mailto:alton.lafferty@outlook.com">alton.lafferty@outlook.com</a></div>
         <div class="ms-PersonaCard-detailLine"><span class="ms-PersonaCard-detailLabel">Work:</span> <a class="ms-Link" href="mailto:alton.lafferty@outlook.com">altonlafferty@contoso.com</a></div>
       </li>
+   </ul>
       <!-- org chart -->
-      <li id="org" class="ms-PersonaCard-actionDetails detail-5">
+   <ul id="detail-6" class="ms-PersonaCard-details">
```
Comment (mikewheaton) [inline src/components/PersonaCard/PersonaCard.html]: Where is `detail-5`? It goes from 4 to 6.

File: src/components/PersonaCard/PersonaCard.Square.html [hunk 3]
```diff
@@ -15,31 +15,39 @@
...
       </li>
-      <li id="video" class="ms-PersonaCard-actionDetails detail-3">
+   </ul>
+   <ul id="detail-3b" class="ms-PersonaCard-details">
+      <li class="ms-PersonaCard-actionDetails">
         <div class="ms-PersonaCard-detailLine"><span class="ms-PersonaCard-detailLabel">Skype:</span> <a class="ms-Link" href="#">Start Skype call</a></div>
       </li>
-      <li id="mail" class="ms-PersonaCard-actionDetails detail-4">
+   </ul>
+   <ul id="detail-4b" class="ms-PersonaCard-details">
```
Comment (mikewheaton) [inline src/components/PersonaCard/PersonaCard.Square.html]: Why the change the `detail-4b` instead of `detail-4` here? Is it to prevent conflicts when two PersonaCards are on the same page? If so, we should find a more foolproof way to avoid conflicts.

File: src/components/PersonaCard/PersonaCard.Square.html [hunk 4]
```diff
@@ -15,31 +15,39 @@
...
   <div class="ms-PersonaCard-actionDetailBox">
-    <ul id="detailList" class="ms-PersonaCard-detailChat">
-      <li id="chat" class="ms-PersonaCard-actionDetails detail-1">
+    <ul id="detail-1b" class="ms-PersonaCard-details">
+      <li  class="ms-PersonaCard-actionDetails">
         <div class="ms-PersonaCard-detailLine"><span class="ms-PersonaCard-detailLabel">Lync:</span> <a class="ms-Link" href="#">Start Lync call</a></div>
       </li>
-      <li id="phone" class="ms-PersonaCard-actionDetails detail-2">
+   </ul>
+   <ul id="detail-2b" class="ms-PersonaCard-details">
```
Comment (mikewheaton) [inline src/components/PersonaCard/PersonaCard.Square.html]: Rather than connecting a data attribute (`data-detailsTargetId="detail-2b"`) to an ID, could we connect two data attributes? It would be best to avoid IDs entirely as they can easily be given names that unknowingly clash with existing styles. Better to stay clear of CSS Land entirely and use data attributes.

**Comment by manndave** [comment]
I'll dig back in to this and make the changes, but it will be a few days.  I'll repost the PR when it is ready.  Thanks.

**Comment by manndave** [comment]
@mikewheaton: I think I got all of your comments addressed.  I also added an example of using the expander class to show details in an action

**Comment by mikewheaton** [comment]
Looks good to me! This is a huge improvement, as we now have a more general way to connect the buttons and content areas (without hard-coding the tabs) and we can have multiple PersonaCards on the same page without the need for unique IDs.

We just need one more review to merge. :)

File: src/components/PersonaCard/PersonaCard.Square.html [hunk 5]
```diff
@@ -15,31 +15,39 @@
...
+   <ul data-Name="mailAction" class="ms-PersonaCard-details">
+      <li class="ms-PersonaCard-actionDetails">
         <div class="ms-PersonaCard-detailLine"><span class="ms-PersonaCard-detailLabel">Personal:</span> <a class="ms-Link" href="mailto:alton.lafferty@outlook.com">alton.lafferty@outlook.com</a></div>
         <div class="ms-PersonaCard-detailLine"><span class="ms-PersonaCard-detailLabel">Work:</span> <a class="ms-Link" href="mailto:alton.lafferty@outlook.com">altonlafferty@contoso.com</a></div>
       </li>
+   </ul>
       <!-- org chart -->
-      <li id="org" class="ms-PersonaCard-actionDetails detail-5">
+   <ul data-Name="orgAction" class="ms-PersonaCard-details ms-PersonaCard-detailsTall ">
+      <li class="ms-PersonaCard-actionDetails>
```
Comment (wdo3650) [inline src/components/PersonaCard/PersonaCard.Square.html]: Missing closing quotes.

File: src/components/PersonaCard/PersonaCard.html [hunk 6]
```diff
@@ -17,31 +17,42 @@
...
+          <div class="ms-PersonaCard-detailExpander"></div>
+          <div class="ms-PersonaCard-detailLine"><span class="ms-PersonaCard-detailLabel">Details</div>
         <div class="ms-PersonaCard-detailLine"><span class="ms-PersonaCard-detailLabel">Personal:</span> <a class="ms-Link" href="mailto:alton.lafferty@outlook.com">alton.lafferty@outlook.com</a></div>
         <div class="ms-PersonaCard-detailLine"><span class="ms-PersonaCard-detailLabel">Work:</span> <a class="ms-Link" href="mailto:alton.lafferty@outlook.com">altonlafferty@contoso.com</a></div>
       </li>
+   </ul>
       <!-- org chart -->
-      <li id="org" class="ms-PersonaCard-actionDetails detail-5">
+   <ul data-Name="orgAction" class="ms-PersonaCard-details ms-PersonaCard-detailsTall ">
+      <li class="ms-PersonaCard-actionDetails>
```
Comment (wdo3650) [inline src/components/PersonaCard/PersonaCard.html]: Missing closing quotes.

File: src/components/PersonaCard/PersonaCard.html [hunk 7]
```diff
@@ -17,31 +17,42 @@
...
-    <ul id="detailList" class="ms-PersonaCard-detailChat">
-      <li id="chat" class="ms-PersonaCard-actionDetails detail-1">
+    <ul data-Name="chatAction" class="ms-PersonaCard-details">
+      <li  class="ms-PersonaCard-actionDetails">
         <div class="ms-PersonaCard-detailLine"><span class="ms-PersonaCard-detailLabel">Lync:</span> <a class="ms-Link" href="#">Start Lync call</a></div>
       </li>
-      <li id="phone" class="ms-PersonaCard-actionDetails detail-2">
+   </ul>
+   <ul data-Name="phoneAction" class="ms-PersonaCard-details">
+      <li class="ms-PersonaCard-actionDetails">
```
Comment (wdo3650) [inline src/components/PersonaCard/PersonaCard.html]: Remove whitespace at the end of the line.

File: src/components/PersonaCard/Jquery.PersonaCard.js [hunk 8]
```diff
@@ -8,50 +8,62 @@
...
-(function ($) {
-  $.fn.PersonaCard = function () {
+(function ($) {    
+    /* Show/Hide the appropriate details for the selected Action */
+    function toggleDetails(target, $personaCard){		
+        if('undefined' === typeof target){
+            var targetIds = [];
+            $('.ms-PersonaCard-action.is-active').each(function(){
+                target = $(this).attr('data-detailsTargetName');
+		        if('undefined' !== typeof target){
```
Comment (wdo3650) [inline src/components/PersonaCard/Jquery.PersonaCard.js]: Indent this line within the function block.

File: src/components/PersonaCard/Jquery.PersonaCard.js [hunk 9]
```diff
@@ -8,50 +8,62 @@
...
+                    targetIds.push(target);
+                }
+            });
+            targetIds.forEach(function(a){toggleDetails(a, $personaCard);});
+        }
+        else{
```
Comment (wdo3650) [inline src/components/PersonaCard/Jquery.PersonaCard.js]: Coding convention thing, move this 'else' up a line, directly after the above closing curly brace.

File: src/components/PersonaCard/Jquery.PersonaCard.js [hunk 10]
```diff
@@ -8,50 +8,62 @@
...
+            var actionUrl = $personaCard.find("li[data-detailsTargetName='" + target + "']").attr('data-actionUrl');
+            if("undefined" === typeof actionUrl){
+                var t = $personaCard.find("ul[data-Name='" + target + "']");
```
Comment (wdo3650) [inline src/components/PersonaCard/Jquery.PersonaCard.js]: Favor variables that confer meaning for readability.

File: src/components/PersonaCard/Jquery.PersonaCard.js [hunk 11]
```diff
@@ -8,50 +8,62 @@
...
+                t.parent().children().removeClass('is-active').hide();
+		        t.addClass('is-active').show();
```
Comment (wdo3650) [inline src/components/PersonaCard/Jquery.PersonaCard.js]: Indent this directly below the line above.

File: src/components/PersonaCard/Jquery.PersonaCard.js [hunk 12]
```diff
@@ -8,50 +8,62 @@
  * @param  {jQuery Object}  One or more .ms-PersonaCard components
  * @return {jQuery Object}  The same components (allows for chaining)
  */
-(function ($) {
-  $.fn.PersonaCard = function () {
+(function ($) {    
+    /* Show/Hide the appropriate details for the selected Action */
+    function toggleDetails(target, $personaCard){
```
Comment (wdo3650) [inline src/components/PersonaCard/Jquery.PersonaCard.js]: Coding convention thing, use 2 spaces instead of 4.

File: src/components/PersonaCard/Jquery.PersonaCard.js [hunk 13]
```diff
@@ -8,50 +8,62 @@
...
+                t.parent().children().removeClass('is-active').hide();
+		        t.addClass('is-active').show();
+            }
+            else{
+                document.location.href=actionUrl;
+            }
+        }
+	}    
+  
+    $.fn.PersonaCard = function () {
```
Comment (wdo3650) [inline src/components/PersonaCard/Jquery.PersonaCard.js]: Reduce the indent.

**Comment by wdo3650** [comment]
@manndave Removing the id's is a great idea and something definitely necessary. Thank you for looking out on this and taking steps to fix this issue. As you may know Fabric will be removing its dependency on JQuery in favor of vanilla Javascript implementations, which we are currently working on. This will directly affect the JQuery in this PR. 

The Fabric team will definitely use the ideas and work from this PR in the vanilla javascript version. If you like you can still go ahead and respond to the code review comments, keep in mind JQuery will be removed in the next major release.

File: src/components/PersonaCard/Jquery.PersonaCard.js [hunk 14]
```diff
@@ -8,50 +8,62 @@
...
+      }
+      
       /** Toggle more details. */
       $personaCard.on('click', '.ms-PersonaCard-detailExpander', function() {
         $(this).parent('.ms-PersonaCard-actionDetails').toggleClass('is-collapsed');
       });
 
     });
 
+
```
Comment (wdo3650) [inline src/components/PersonaCard/Jquery.PersonaCard.js]: Remove whitespace

**Comment by wdo3650** [comment]
@manndave the animation of the action detail box growing and contracting when switching between action details seems to be missing.

**Comment by manndave** [comment]
At this point, it probably makes the most sense not to pursue this PR since the team is switching to vanilla-JS.  If you want me to do the vanilla-JS version of this let me know and I'll do a new PR

**Comment by mikewheaton** [comment]
Thanks for your understanding here, @manndave. We're actively working on 3.0.0 and want to provide much more robust, production-ready JavaScript in the future. Let's close this pull request and then make sure we take this thinking into the new PersonaCard.
